### ISSUE TITLE
Can't apply on Server Core (WMI Issues)

### ISSUE THREAD

**Issue by DavidRawling** [J0F3/cPowerPlan#1]
During configuration of a Server Core host, if a PowerPlan is selected configuration of that component will fail:

```
VERBOSE: [ZZZ-HOST1]:                            [[cPowerPlan]HighPerformance] Perform operation 'Enumerate CimInstances' with following parameters, ''namespaceName' = root\cimv2\power,'className' = win32_PowerPlan'.
Provider load failure
    + CategoryInfo          : NotSpecified: (root\cimv2\power:win32_PowerPlan:) [], CimException
    + FullyQualifiedErrorId : HRESULT 0x80041013,Microsoft.Management.Infrastructure.CimCmdlets.GetCimInstanceCommand
    + PSComputerName        : localhost

VERBOSE: [ZZZ-HOST1]:                            [[cPowerPlan]HighPerformance] Operation 'Enumerate CimInstances' complete.
VERBOSE: [ZZZ-HOST1]: LCM:  [ End    Test     ]  [[cPowerPlan]HighPerformance]  in 0.0430 seconds.
The PowerShell DSC resource '[cPowerPlan]HighPerformance' with SourceInfo 'D:\Sources\PowerPlan_HighPerf.ps1::16::9::cPowerPlan' threw one or more non-terminating errors while running the
Test-TargetResource functionality. These errors are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
```

This appears to be due to missing WMI classes on Server Core environments.

https://vniklas.djungeln.se/2014/12/02/get-and-set-powerplan-on-windows-hyper-v-2012-r2-core/ suggests an approach using powercfg to get and set. https://p0w3rsh3ll.wordpress.com/2014/12/28/fixing-a-wmi-provider-load-failure/ suggests copying missing files.

I don't know that the fix needs to be in cPowerPlan, but it's worth documenting as a limitation, at least.

**Comment by DavidRawling** [comment]
If someone does need the workaround, try this Script resource as a temp fix:

```
$DSCPowerPlan = "High Performance"

Script PowerPlan
{
  SetScript = 
  {
    # Get the list of Power Plans
    $PowerCfg = (Invoke-Command { PowerCfg -l })
    $PowerCfg = $PowerCfg[3..$PowerCfg.Count]

    # Locate the High Performance plan
    $PlanItem = ($PowerCfg | Select-String -Pattern $Using:DSCPowerPlan -CaseSensitive:$False -SimpleMatch).ToString()
    $IX0 = $PlanItem.IndexOf(":")
    $IX1 = $PlanItem.IndexOf(" ", $IX0+1)
    $IX2 = $PlanItem.IndexOf(" ", $IX1+1)
    $PlanGUID = $PlanItem.SubString($IX1+1, $IX2-$IX1-1)

    PowerCfg -s $PlanGUID
  }

  GetScript = 
  {
    # Get the list of Power Plans
    $PowerCfg = (Invoke-Command { PowerCfg -l })
    $PowerCfg = $PowerCfg[3..$PowerCfg.Count]

    # Locate the Active plan
    $ActiveItem = ($PowerCfg | Select-String -Pattern "*" -CaseSensitive:$False -SimpleMatch).ToString()
    $IX1 = $ActiveItem.IndexOf("(")
    $IX2 = $ActiveItem.IndexOf(")")
    $ActivePlan = $ActiveItem.SubString($IX1+1, $IX2-$IX1-1)

    return @{ 'Result' = "Power Plan: $ActivePlan" }
  }

  TestScript = 
  {
    # Get the list of Power Plans
    $PowerCfg = (Invoke-Command { PowerCfg -l })
    $PowerCfg = $PowerCfg[3..$PowerCfg.Count]

    # Locate the High Performance plan
    $PlanItem = ($PowerCfg | Select-String -Pattern $Using:DSCPowerPlan -CaseSensitive:$False -SimpleMatch).ToString()
    $IX0 = $PlanItem.IndexOf(":")
    $IX1 = $PlanItem.IndexOf(" ", $IX0+1)
    $IX2 = $PlanItem.IndexOf(" ", $IX1+1)
    $PlanGUID = $PlanItem.SubString($IX1+1, $IX2-$IX1-1)

    # Locate the Active plan
    $ActiveItem = ($PowerCfg | Select-String -Pattern "*" -CaseSensitive:$False -SimpleMatch).ToString()
    $IX0 = $ActiveItem.IndexOf(":")
    $IX1 = $ActiveItem.IndexOf(" ", $IX0+1)
    $IX2 = $ActiveItem.IndexOf(" ", $IX1+1)
    $ActiveGUID = $ActiveItem.SubString($IX1+1, $IX2-$IX1-1)

    Return ($ActiveGUID.ToUpper() -eq $PlanGUID.ToUpper())
  }
}
```

**Comment by J0F3** [comment]
Hi David,

Sorry for my delay in answer! 

I am much appreciate your input. You are absolutely right. The WMI class is not available in core server and in nano server.

Actually for core server I have filed a bug on conenct and therefore the WMI class is now available again on Server 2016. 
https://connect.microsoft.com/WindowsServer/feedback/details/2397595/wmi-class-win32-powerplan-not-available-on-server-core-installation

Hower for with Nano server the issue with the missing WMI class exists again. So I decided to rewrite the resource and implemented a differnt set and get mechanism which is partially based on your posted workaround. 

I have just merge my changes in the master branch of this repository (#2) and I have also published a new version (1.0.1.0) of the Module to the PowerShell Gallery.

As the resource works now on core and nano server I will close this issue now but feel free to get back to me if you see any other or new issues with the resource.

Thanks
Jonas
